components/filesystem/include/filesystem/filesystem.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "core/platform/context.hpp"
+#include "core/util/logging.hpp"
+
+namespace vkb
+{
+namespace fs
+{
+struct FileStat
+{
+	bool   is_file;
+	bool   is_directory;
+	size_t size;
+};
+
+class FileSystem
+{
+  public:
+	FileSystem()          = default;
+	virtual ~FileSystem() = default;
+
+	virtual FileStat             stat_file(const std::string &path)                                    = 0;
+	virtual bool                 is_file(const std::string &path)                                      = 0;
+	virtual bool                 is_directory(const std::string &path)                                 = 0;
+	virtual bool                 exists(const std::string &path)                                       = 0;
+	virtual bool                 create_directory(const std::string &path)                             = 0;
+	virtual std::vector<uint8_t> read_chunk(const std::string &path, size_t start, size_t offset)      = 0;
+	virtual void                 write_file(const std::string &path, const std::vector<uint8_t> &data) = 0;
+
+	void write_file(const std::string &path, const std::string &data);
+
+	// Read the entire file into a string
+	std::string read_file(const std::string &path);
+
+	// Read the entire file into a vector of bytes
+	template <typename T = uint8_t>
+	std::vector<T> read_binary_file(const std::string &path)
```

Comment:
You might consider a function along the lines of

```c++
template<typename T = uint8_t>
void with_file_contents(const std::string& filename, const std::function<void(size_t, const T*)>& handler);
```

In my experience, most applications only need the asset data for the scope of the call where it's read.  While `read_binary_file` returns a `std::vector` and thus _always_ allocates heap space, the implementation of something like `with_file_contents` is free to give a pointer to a memory mapped file, or something like the result of an `AAsset_getBuffer(...)` call on Android, which can lower the amount of actual heap space used in the application.  Obviously anything going into a staging buffer will consume host memory anyway, but at least this would potentially cut that in half.

Reply:
I spun of a filesystem PR to simplify this one. I will move this comment there tomorrow